src/client/dfuse/ops/lookup.c

```diff
@@ -21,95 +21,102 @@
+	DFUSE_TRA_INFO(inode, "Inserting inode %lu", entry.ino);
 
-	desc->ie->stat = out->stat;
-	DFUSE_TRA_UP(desc->ie, fs_handle, "inode");
 	rlink = d_hash_rec_find_insert(&fs_handle->inode_ht,
-				       &desc->ie->stat.st_ino,
-				       sizeof(desc->ie->stat.st_ino),
-				       &desc->ie->ie_htl);
-
-	if (rlink == &desc->ie->ie_htl) {
-		desc->ie = NULL;
-		keep_ref = true;
-	} else {
+				       &inode->stat.st_ino,
+				       sizeof(inode->stat.st_ino),
+				       &inode->ie_htl);
+
+	if (rlink != &inode->ie_htl) {
 		/* The lookup has resulted in an existing file, so reuse that
 		 * entry, drop the inode in the lookup descriptor and do not
 		 * keep a reference on the parent.
 		 * Note that this function will be called with a reference on
 		 * the parent anyway, so keep that one, but drop one in the call
 		 * to ie_close().
 		 */
-		atomic_fetch_sub(&desc->ie->ie_ref, 1);
-		keep_ref = true;
-		ie_close(fs_handle, desc->ie);
+		atomic_fetch_sub(&inode->ie_ref, 1);
+
+		ie_close(fs_handle, inode);
 	}
 
-	DFUSE_REPLY_ENTRY(request, entry);
-	dfuse_da_release(desc->da, desc);
-	return keep_ref;
-out:
-	DFUSE_REPLY_ERR(request, request->rc);
-	dfuse_da_release(desc->da, desc);
-	return false;
+	DFUSE_REPLY_ENTRY(req, entry);
 }
 
-static const struct dfuse_request_api api = {
-	.on_result	= dfuse_entry_cb,
-};
-
 void
 dfuse_cb_lookup(fuse_req_t req, fuse_ino_t parent, const char *name)
 {
 	struct dfuse_projection_info	*fs_handle = fuse_req_userdata(req);
-	struct TYPE_NAME		*desc = NULL;
+	struct dfuse_inode_entry	*inode = NULL;
+	struct dfuse_inode_entry	*parent_inode;
+	mode_t				mode;
+	d_list_t			*rlink;
 	int rc;
 
 	DFUSE_TRA_INFO(fs_handle, "Parent:%lu '%s'", parent, name);
-	DFUSE_REQ_INIT_REQ(desc, fs_handle, api, req, rc);
-	if (rc)
-		D_GOTO(err, rc);
 
-	DFUSE_TRA_INFO(desc, "ie %p", &desc->ie);
+	rlink = d_hash_rec_find(&fs_handle->inode_ht, &parent, sizeof(parent));
+	if (!rlink) {
+		DFUSE_TRA_ERROR(fs_handle, "Failed to find inode %lu",
+				parent);
+		D_GOTO(err, rc = ENOENT);
+	}
 
-	desc->request.ir_inode_num = parent;
+	parent_inode = container_of(rlink, struct dfuse_inode_entry, ie_htl);
 
-	strncpy(desc->ie->name, name, NAME_MAX);
-	desc->ie->parent = parent;
-	desc->da = fs_handle->lookup_da;
+	DFUSE_TRA_INFO(parent_inode, "parent");
 
-	rc = dfuse_fs_send(&desc->request);
-	if (rc != 0)
-		D_GOTO(err, 0);
+	D_ALLOC_PTR(inode);
+	if (!inode) {
+		D_GOTO(out_decref, rc = ENOMEM);
+	}
+	inode->parent = parent;
+
+	rc = dfs_lookup_rel(fs_handle->fsh_dfs, parent_inode->obj, name,
+			    O_RDONLY, &inode->obj, &mode);
+	if (rc != -DER_SUCCESS) {
+		DFUSE_TRA_INFO(fs_handle, "dfs_lookup() failed: %p %d",
+			       fs_handle->fsh_dfs, rc);
+		if (rc == -DER_NONEXIST) {
+			D_GOTO(out_decref, rc = ENOENT);
+		} else {
+			D_GOTO(out_decref, rc = EIO);
+		}
+	}
+
+	strncpy(inode->name, name, NAME_MAX);
+	atomic_fetch_add(&inode->ie_ref, 1);
+
+	rc = dfs_ostat(fs_handle->fsh_dfs, inode->obj, &inode->stat);
```

Comment:
why do we ostat after a lookup?

Reply:
Same as mkdir, the reply to lookup is fuse_reply_entry() which passes a stat call into the kernel.